### PR TITLE
1d benchmark

### PR DESCRIPTION
**Pull request by arthbous**
Everything is coded but it is not converging at all...

### CODE REVIEW

**Comment by arthbous** [comment]
* It converges slowly but surely with dt=0.01
* It gets weird with dt=0.1, because of to many mesh refinement I have an error that occurs.

If you could  "make pnp_1dbench && ./pnp_1dbench" with dt=0.1 and let me know if you also get an error that would be great.

I will add the energy and start putting the results with dt=0.01 for Wenxiao.

**Comment by maximilianmetti** [comment]
- The refinement is going to be weird for the time-dependent case.  So it might be reasonable to set the ``entropy_tol`` to be a little larger.  Running with ``entropy_tol = 1.0e-5`` I was able to solve to ``t=1.0`` with ``dt=0.1`` and also run a single step with ``dt=1.01``.
- I *did* change the ``free``ing of ``solu_fasp`` and ``A_fasp_bsr``, which was included on the commit I made.
- From what I see, I'm not sure that the solution at the previous time-step is getting updated correctly... Maybe this is why the solution doesn't look like it's changing at all...  I need to take a closer look (I'm a bit tired to really understand which variables are which).

**Comment by arthbous** [comment]
* I like where you put the free fasp. There were still some free fasp that I had put that I removed, I just kept yours.
* I separated the files into two files: pnp_adapt in which I keep the mesh adapted at each time step and pnp_adapt2 which is like before the mesh adapt is reset at each time step. I think it makes more sens to no reset the adapted mesh.
* I added some printing in data.txt and data2.txt

**Comment by maximilianmetti** [comment]
cool.  I think keeping the mesh refinement will be good...  we just need to make sure we don't over-refine since we never coarsen.

**Comment by maximilianmetti** [comment]
I took a look at the 1d-benchmark.  I made some changes and it looks like its working now.  I have PR #41  to merge into this branch with details there.

One comment about naming conventions for variables: it's better to avoid names like ``t``, ``dt``, or ``tf`` since they don't hold any meaning to new users, also they're hard to search for.  I always think it's better to be *overly descriptive* with names to avoid ambiguity and make their functionality clear.  Please review the PR.

**Comment by maximilianmetti** [comment]
at the end of the simulation, some memory isn't freed... : /
``[tianhe2.attlocal.net:80481] [[13956,1],0]-[[13956,0],0] mca_oob_tcp_msg_recv: readv failed: Operation timed out (60)``

**Comment by arthbous** [comment]
I added the computation of phi in term of cation and anion, see "benchmarks/1d-benchmark/pnp_adapt3.cpp", "benchmarks/1d-benchmark/poisson.ufl". 
To test it just run **make pnp_1dbench3 && ./pnp_1dbench3"

**Comment by maximilianmetti** [comment]
The solve for \phi should help the initial guess (perhaps keep the Newton solver from doing weird stuff in the beginning).

In reference to small epsilon, the dimensionless form of the PNP equations suggest that epsilon is the square of the Debye length over the reference length of the domain.  Since the Debye length approximates the thickness of the boundary layer, that gives an idea of how fine the mesh should be; the mesh should refine to fit a couple elements in the boundary layer which is about 0.03.  I'm guessing it needs to refine the mesh to have a diameter of roughly 0.01 units before capturing the solution accurately.  The solver should still work since we got it working for eps = 1e-8 before (steady state).

**Comment by arthbous** [comment]
Yes, the problem is when I start with a fine mesh it takes way to long to compute like 30 minutes per gmres. 
And the refinement nothing happens, I am thinking that maybe refining on the values of phi might not be the best criteria.

**Comment by maximilianmetti** [comment]
The solver should work even if the mesh is coarse; it will just give a
crappy approximation.  If you start with a coarse mesh, it should refine on
its own to resolve the boundary layers.  What happens when you do a coarse
solve?  Does the solver fail to converge or throw a segfault?

**Comment by arthbous** [comment]
The solver gives an nan solutions, gmres does not solve anything

**Comment by maximilianmetti** [comment]
WTF???  If you make an analogous problem using the steady state solver
``benchmarks/PNP/pnp_adapt.cpp``, do you also get ``NaN``s?  That could
help us discover a bug! : D

Gotta love finding bugs brah!

**Comment by arthbous** [comment]
it seems to be the same problem, the solver does not converges once we do some refinement.

Also it seems to make no difference we can take any epsilon nothing changes.

**Comment by maximilianmetti** [comment]
can you send me the problem setup so i can reproduce?

**Comment by arthbous** [comment]
You can do with the master branch in benchmarks/PNP/pnp_adapt.cpp take permitivity = 0.001 and entropy_tol = 1.0e-5.

**Comment by maximilianmetti** [comment]
i just ran that.  i get a solution, just fine.  i used eps = 1e-4

**Comment by arthbous** [comment]
Does it refine ?

**Comment by maximilianmetti** [comment]
it refines

**Comment by arthbous** [comment]
ANd if you run the same with the 1d-benchmark branch, where the entropy is different?

**Comment by maximilianmetti** [comment]
i am on the master branch...

**Comment by arthbous** [comment]
Yup if you swith branch and run the same test. Also on the master branch when I run the test I am stuck..it does not converges.

**Comment by arthbous** [comment]
Could you make push the test you did to master? And I will download it exactly and run it.

**Comment by maximilianmetti** [comment]
bingo...
screen shot 2016-01-04 at 2 37 48 pm

**Comment by maximilianmetti** [comment]
well, it seems we solved one problem, only to encounter another...
I've been running the script for a while and the FEniCS functions make the mesh adaptivity take a very, very long time.  I'm letting this run and will look at it tomorrow, but I'd like to suggest  that we
- try using the master version of ``domains.cpp`` and see if this issue crops up (with small ``adapt_tol``)
- try using ``epsilon * grad(phi)`` as the error indicator so that ``entropy_tol`` scales better, as it will measure the electrical intensity... Note that there will be really steep boundary layers that occur for small ``epsilon``, so this scaling will avoid over-refinement.

I'll try to reproduce the singular matrix, but I've been waiting a while and still haven't seen it.  I'm setting a higher print_level in ``src/gradient_recovery_bsr.dat`` for help to debug

**Comment by maximilianmetti** [comment]
I saw the error, but wasn't able to figure it out... I'm guessing that we're using some wrong solver in FASP, since it reports many nearly singular matrices, it seems to be that it could be related to some aspect of the solver that is under-the-hood....  I'm going to try playing with different solvers, but am not sure i'll get far.

File: src/domains.cpp [hunk 1]
```diff
@@ -153,15 +155,15 @@ unsigned int check_local_entropy (dolfin::Function *cation,
   status = fasp_solver_dbsr_krylov_diag(&adaptA_fasp_bsr, &b_fasp, &solu_fasp, &itpar);
```
Comment (maximilianmetti) [inline src/domains.cpp]: To get this working, replace this line with
```
status = fasp_solver_dcsr_krylov_diag(&A_fasp, &b_fasp, &solu_fasp, &itpar);
```
seems that the problem was with using a block solver.

**Comment by maximilianmetti** [comment]
just a heads up: we're totally over-solving the problem.  The Newton solver cannot reduce the solution by a factor of 1E-8 when delta t is small.

File: benchmarks/1d-benchmark/pnp_adaptive.cpp [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+  Constant Dn(coeff_par.anion_diffusivity);
+  Constant qp(coeff_par.cation_valency);
+  Constant qn(coeff_par.anion_valency);
+  Constant C_dt(time_step_size);
+  Constant cat_alpha(coeff_par.cation_diffusivity*time_step_size);
+  Constant an_alpha(coeff_par.anion_diffusivity*time_step_size);
+  Constant C1(1.0);
+  Constant zero(0.0);
+
+  for (double t = 0; t < final_time; t += time_step_size) {
```
Comment (maximilianmetti) [inline benchmarks/1d-benchmark/pnp_adaptive.cpp]: this needs to become
```for (double t = time_step_size; ... ) {```
the initial guess should serve as ``t=0.0``.

**Comment by arthbous** [comment]
We should also change to a dcsr solver in Master since we only one unknown, we should NOT use the block solver.

**Comment by maximilianmetti** [comment]
Absolutely. Apparently, using a block solver is bad when doing aggressive refinement.

**Comment by arthbous** [comment]
* I added  ``check_local_entropy`` and ``check_electric_field`` in the same file.

* I also added **1d-benchmarkbis/**, by starting from the new **PNP/pnp_adaptive.cpp** => I get the same errors.

Remark: The segmentation fault that only appears after the conversion to FASP is very dependent on the same of the mesh.

**Comment by arthbous** [comment]
@maximilianmetti  I made a lot of changes. 

* Now it is PNP_time/ and pnp_time.cpp
* I moved some files in include
* I added a ufl folder

Please review this branch and merge it

**Pull request merged**